Type hints for lib module

umak1106 opened the pull request:
Fixes #

Changes made in this Pull Request:
 - 


PR Checklist
------------
 - [ ] Tests?
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [ ] Issue raised/referenced?

umak1106 committed:
Type hints for mdamath.py

pep8speaks commented:
Hello @umak1106! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`package/MDAnalysis/__init__.py`](https://github.com/MDAnalysis/mdanalysis/blob/afbf3cabd43ec0b5edd2c2e5d0d2a6f4f00c76d9/package/MDAnalysis/__init__.py):

> [Line 182:80](https://github.com/MDAnalysis/mdanalysis/blob/afbf3cabd43ec0b5edd2c2e5d0d2a6f4f00c76d9/package/MDAnalysis/__init__.py#L182): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)



* In the file [`package/MDAnalysis/lib/mdamath.py`](https://github.com/MDAnalysis/mdanalysis/blob/afbf3cabd43ec0b5edd2c2e5d0d2a6f4f00c76d9/package/MDAnalysis/lib/mdamath.py):

> [Line 164:80](https://github.com/MDAnalysis/mdanalysis/blob/afbf3cabd43ec0b5edd2c2e5d0d2a6f4f00c76d9/package/MDAnalysis/lib/mdamath.py#L164): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 249:80](https://github.com/MDAnalysis/mdanalysis/blob/afbf3cabd43ec0b5edd2c2e5d0d2a6f4f00c76d9/package/MDAnalysis/lib/mdamath.py#L249): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 312:80](https://github.com/MDAnalysis/mdanalysis/blob/afbf3cabd43ec0b5edd2c2e5d0d2a6f4f00c76d9/package/MDAnalysis/lib/mdamath.py#L312): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (98 > 79 characters)

* In the file [`package/MDAnalysis/lib/pkdtree.py`](https://github.com/MDAnalysis/mdanalysis/blob/afbf3cabd43ec0b5edd2c2e5d0d2a6f4f00c76d9/package/MDAnalysis/lib/pkdtree.py):

> [Line 101:80](https://github.com/MDAnalysis/mdanalysis/blob/afbf3cabd43ec0b5edd2c2e5d0d2a6f4f00c76d9/package/MDAnalysis/lib/pkdtree.py#L101): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)



##### Comment last updated at 2022-09-12 07:29:38 UTC

umak1106 commented:
Regarding the stp() function if i set the arguments as arraylike i get this error
Argument 1 to "cross" has incompatible type "Union[_SupportsArray[dtype[Any]], _NestedSequence[_SupportsArray[dtype[Any]]], bool, int, float, complex, str, bytes, _NestedSequence[Union[bool, int, float, complex, str, bytes]]]"; expected "Union[_SupportsArray[dtype[bool_]], _NestedSequence[_SupportsArray[dtype[bool_]]], bool, _NestedSequence[bool]]"

And if i set it as NDArray I still get an error
Argument 1 to "stp" has incompatible type "Union[_SupportsArray[dtype[Any]], _NestedSequence[_SupportsArray[dtype[Any]]], bool, int, float, complex, str, bytes, _NestedSequence[Union[bool, int, float, complex, str, bytes]]]"; expected "ndarray[Any, dtype[Any]]"

The same error above shows  for normal()

umak1106 committed:
fix errors in mdamath.py

umak1106 commented:
actually typing module is introduced in numpy from 1.20.0 and the gh-ci.yaml file has numpy = 1.19.0 which is why i think the tests are failing

jbarnoud commented:
@IAlibay plans on bumping the minimal version of numpy to 1.20.0. It means we'll have to wait for the bump to happen before we merge, but you can proceed and ignore the failing pipeline.

jbarnoud commented on package/MDAnalysis/lib/mdamath.py:
```diff
@@ -90,7 +92,7 @@ def norm(v):
 
 
-def normal(vec1, vec2):
+def normal(vec1: npt.NDArray, vec2: npt.NDArray) -> npt.NDArray:
```
Here, the inputs can be `ArrayLike`, I think.

The idea is to be loose on the input to accommodate everything that is valid, but precise on the output.

umak1106 replied:
Ok I'll make the changes

jbarnoud commented on package/MDAnalysis/lib/mdamath.py:
```diff
@@ -110,7 +112,7 @@ def normal(vec1, vec2):
 
 
-def pdot(a, b):
+def pdot(a: npt.NDArray, b: npt.NDArray) -> npt.NDArray:
```
Here too, I think the input can be `ArrayLike`. It should be true for other functions as well.

umak1106 replied:
Actually the doctoring said ndarray so I did that but I'll make the changes.
After making the changes I get error on that line

jbarnoud replied:
Automatic tools may tell you that, indeed, because it is how it is use in the examples it seen. But if you read the code, you see that `a` and `b`  are passed to the `np.einsum`function that expects an `ArrayLike` there.

jbarnoud reviewed (commented):
Do you still have the error you described with `normal` and `stp`? I did not get them when testing locally.

I only have small comments about the typing itself. In general, we want to be broad when it comes to inputs and as specific as possible for the outputs. This is because we do not know what we will get as inputs and will accept anything that can work, but (most of the time) we know exactly what we will output.

You should remove the files you annotate from those ignored by mypy so that the CI actually runs mypy on the files you edit.

hmacdope commented:
@umak1106 just a quick note, I am adding type hints to a lot of `distances.py` as part of #3730, you can probably leave that one to me for now.

umak1106 commented:
ok

umak1106 committed:
changed input from NDarray to arraylike

umak1106 committed:
Added type annotations to init.py 

Added type annotations to init to avoid mypy from raising errors when other modules are type checked .

umak1106 committed:
Allowing mypy to type check lib module

jbarnoud commented on package/MDAnalysis/__init__.py:
```diff
@@ -167,18 +167,18 @@
-_PARSER_HINTS = {}
-_SELECTION_WRITERS = {}
-_CONVERTERS = {}
+_READERS = {}  # type: dict
```
Is there a reason for the different style of annotation? If not, I'd rather have the style consistent:

```python
_READERS: Dict = {}
```

Also, the annotation can be more specific:

```python
_READERS: Dict[str, ProtoReader] = {}
```

umak1106 replied:
Ok I'll make the changes

umak1106 replied:
I made the changes like you suggested . But the tests are failing

jbarnoud replied:
I looked at it and it looks like the issue is caused by the imports in `__init__.py`.

I tried two ways of fixing the problem:
* I guarded the import with `TYPE_CHECKING`, but then I needed to reference the type with forward references (using the name of the types as strings instead of just providing the types)
* providing the types in a stub file (see <https://mypy.readthedocs.io/en/stable/stubs.html>)

Both methods seem to work in the sense that nowhere mypy complains about not having annotations for these variables. However, I know the annotations are wrong, yet I do not get any errors about them. So I would suggest keeping it simple for now and just annotating them as `Dict`. It will not be as precise as it should, but at least it won't be wrong.

umak1106 committed:
Update changes in init.py

umak1106 committed:
Update __init__.py

umak1106 committed:
type hints for pkdtree

umak1106 commented:
Errors am facing :
MDAnalysis/lib/pkdtree.py:141: error: Incompatible types in assignment (expression has type "Optional[float]", variable has type "None")
MDAnalysis/lib/pkdtree.py:184: error: Unsupported operand types for > ("float" and "None")
MDAnalysis/lib/pkdtree.py:234: error: Unsupported operand types for > ("float" and "None")
MDAnalysis/lib/pkdtree.py:290: error: Unsupported operand types for > ("float" and "None")

umak1106 commented on package/MDAnalysis/lib/pkdtree.py:
```diff
@@ -202,7 +204,7 @@ def search(self, centers, radius):
         return self._indices
 
-    def get_indices(self):
+    def get_indices(self) -> npt.NDArray:
```
according to the docstring self._indices should be a list but doesn't np.asarray return NDArray

jbarnoud replied:
Looks like the docstring is out of date, indeed. Could you fix it?

umak1106 commented on package/MDAnalysis/lib/pkdtree.py:
```diff
@@ -245,7 +247,7 @@ def search_pairs(self, radius):
         return pairs
 
-    def search_tree(self, centers, radius):
+    def search_tree(self, centers: npt.ArrayLike, radius: float) -> npt.NDArray:
```
Same as above

jbarnoud replied:
Here the docstring and the code seem to match, don't they? `pair` is an array is I recall correctly.

umak1106 replied:
Yes I misread . The doctoring mentions array .

jbarnoud commented on package/MDAnalysis/lib/pkdtree.py:
```diff
@@ -95,7 +97,7 @@ def pbc(self):
         return self.box is not None
 
-    def set_coords(self, coords, cutoff=None):
+    def set_coords(self, coords: npt.ArrayLike, cutoff: Union[float, None] = None) -> None:
```
```suggestion
    def set_coords(self, coords: npt.ArrayLike, cutoff: Optional[float] = None) -> None:
```

jbarnoud reviewed (commented):
Regarding the errors you report:
* for the first one, you probably should type the instance attribute. See https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html#classes
* for the other errors, we assume that cutoff is set if pbc is True. It holds with the logic of how we expect the class to be used, but it can break, which is what mypy is warning us about. You may need to add a few checks, here and there, to make sure that cutoff is not None when pbc is True.

umak1106 committed:
Fix all errros in pkdtree.py

jbarnoud commented on package/MDAnalysis/lib/pkdtree.py:
```diff
@@ -176,7 +178,8 @@ def search(self, centers, radius):
         if centers.shape == (self.dim, ):
             centers = centers.reshape((1, self.dim))
-
+        if self.cutoff is None:
```
`self.cutoff` can be `None` is `self.pbc` is `False`. It is only if `self.pbc == True` that we assume we have a value for the cutoff.

umak1106 replied:
If i use 
if self.pbc and (self.cutoff is None) :
raise Valuerror("Provide cutoff")

mypy still raises this error 
Unsupported operand types for > ("float" and "None")
Left operand is of type "Optional[float]"

jbarnoud replied:
Can you push the code with the conditional moved? This way I can comment on the code you have rather than the code I imagine you may have. Also, it will let us see if the tests, other than mypy, pass.

jbarnoud replied:
This remains outstanding. It looks like it is the latest obstacle in terms of type annotations.

jbarnoud replied:
I did not realise you did push your latest change. Here is a possible solution. I did not test it, but it might work:

```python
if self.pbc:
    if self.cutoff is None:
        raise ValueError("Cutoff needs to be provided when working with PBC.")
    cutoff: float = self.cutoff
    # From here and until the end if the if self.pbc block, use cutoff instead of self.cutoff
```

I haven't tested that solution, but I expect that mypy will figure out that `cutoff` can only be a float.

umak1106 replied:
I tried that but it still raises the same error

jbarnoud replied:
I am confused. I tried on a small example and it works perfectly fine:

```python
from typing import Optional

def test(pbc: bool, cutoff: Optional[float]):
    if pbc:
        #if cutoff is None:
        #    raise ValueError
        if cutoff < 3:
            print('Hello')
    print('Good bye')

test(False, 2)  # OK, cutoff is just ignored
test(False, None)  # OK
test(True, 2)  # OK
test(True, None)  # Not OK
```

With the `if cutoff is None` commented, I get the same mypy error:

```
test.py:7: error: Unsupported operand types for > ("int" and "None")
test.py:7: note: Left operand is of type "Optional[float]"
```

If it is uncommented, I do not get any errors.

umak1106 replied:
Done

jbarnoud commented on package/MDAnalysis/lib/pkdtree.py:
```diff
@@ -176,7 +178,8 @@ def search(self, centers, radius):
             centers = centers.reshape((1, self.dim))
-
+        if self.cutoff is None:
+            raise RuntimeError("Provide cutoff distance")
```
I *think* a `ValueError` would be more appropriate here.

jbarnoud reviewed (changes requested):
Last few small things:

* there is a small misunderstanding with the test about cutoff. I explained this in a comment at one of the places you added the test but it is true for all the places where you have that test.
* there seems to be an issue with using `npt.NDArray` and numpy 1.20.0. See #3748. For the time being, can you use `np.ndarray` instead so we can move forward?

umak1106 committed:
Chage npt.NDArray to np.ndarray

jbarnoud commented:
It looks like you are still using `npt.NDArray` in pkdtree.

umak1106 committed:
Update pkdtree.py

umak1106 committed:
Update pkdtree.py

umak1106 commented:
> It looks like you are still using `npt.NDArray` in pkdtree.

Yes actually after i faced mypy errors there i didn't push my code

umak1106 committed:
Update pkdtree.py

umak1106 committed:
Merge branch 'MDAnalysis:develop' into patch-2

umak1106 committed:
Update NeighborSearch.py

jbarnoud commented on package/MDAnalysis/lib/pkdtree.py:
```diff
@@ -127,22 +130,22 @@ def set_coords(self, coords, cutoff=None):
 
         if self.pbc:
-            self.cutoff = cutoff
+            if self.cutoff is None:
```
Here, the original test is not about `self.cutoff`, it is instead about `cutoff` as provided as an argument to the function. This seems to be why the tests fail.

```suggestion
            if cutoff is None:
```

It is subtle and got me puzzled for a while!

jbarnoud commented on package/MDAnalysis/lib/pkdtree.py:
```diff
@@ -127,22 +130,22 @@ def set_coords(self, coords, cutoff=None):
 
         """
         # If no cutoff distance is provided but PBC aware
```
Move the comment with the code. Now we have an orphan comment that will confuse future contributors and maintainers.

jbarnoud commented on package/MDAnalysis/lib/pkdtree.py:
```diff
@@ -127,22 +130,22 @@ def set_coords(self, coords, cutoff=None):
         if self.pbc:
-            self.cutoff = cutoff
+            if self.cutoff is None:
+                raise ValueError("Cutoff needs to be provided "
+                                 "when working with PBC.")
```
In this specific instance, because you move an existing exception raising, it would be good to raise the same exception as what you removed:

```suggestion
                raise RuntimeError('Provide a cutoff distance'
                               ' with tree.set_coords(...)'
```

umak1106 replied:
Done

jbarnoud commented on package/MDAnalysis/lib/pkdtree.py:
```diff
@@ -127,22 +130,22 @@ def set_coords(self, coords, cutoff=None):
         coords = np.asarray(coords, dtype=np.float32)
 
         if self.pbc:
-            self.cutoff = cutoff
```
This line is extremely important and cannot be removed! This line actually sets the cutoff in the object; without it, `self.cutoff` will always be `None`. The line needs to be run if the exception about the cutoff is not raised.

jbarnoud commented on package/MDAnalysis/lib/NeighborSearch.py:
```diff
@@ -102,7 +109,7 @@ def search(self, atoms, radius, level='A'):
         return self._index2level(unique_idx, level)
 
-    def _index2level(self, indices, level):
+    def _index2level(self, indices: List[int], level: str) -> None:
```
```suggestion
    def _index2level(self, indices: List[int], level: str) -> Union[AtomGroup, ResidueGroup, SegmentGroup]:
```

jbarnoud reviewed (changes requested):
I made some suggestions to fix errors that caused either the tests to fails or mypy to complain. After that, all the complaints from mypy are about `distances.py` that have wrong annotations. I'll try and fix them.

jbarnoud commented:
Now that #3787 is merged into develop, the develop branch should be merged in here.

umak1106 committed:
Merge branch 'MDAnalysis:develop' into patch-2

jbarnoud commented:
Wooo! 🎉 Only one mypy error left! With the last few comment addressed we should be good to go 😁

umak1106 committed:
Update NeighborSearch.py

umak1106 commented:
Making those changes in pkdtree raises errors like unresolved reference to self

jbarnoud commented:
> Making those changes in pkdtree raises errors like unresolved reference to self

Can you paste the full error message please? It is difficult to figure out what can go wrong without it.

jbarnoud commented on package/MDAnalysis/__init__.py:
```diff
@@ -167,18 +169,18 @@
+_READERS: Dict = {}
+_READER_HINTS: Dict = {}
+_SINGLEFRAME_WRITERS: Dict = {}
+_MULTIFRAME_WRITERS: Dict = {}
+_PARSERS: Dict = {}
+_PARSER_HINTS: Dict = {}
+_SELECTION_WRITERS: Dict = {}
+_CONVERTERS: Dict = {}
 # Registry of TopologyAttributes
-_TOPOLOGY_ATTRS = {}  # {attrname: cls}
-_TOPOLOGY_TRANSPLANTS = {}   # {name: [attrname, method, transplant class]}
-_TOPOLOGY_ATTRNAMES = {}  # {lower case name w/o _ : name}
+_TOPOLOGY_ATTRS: Dict = {}   # {attrname: cls}
+_TOPOLOGY_TRANSPLANTS: Dict = {}   # {name: [attrname, method, transplant class]}
+_TOPOLOGY_ATTRNAMES: Dict = {}   # {lower case name w/o _ : name}
```
@umak1106 It would be great to have these lines in a separate PR so we can work at annotating other parts of the code without triggering errors here.

umak1106 replied:
ok mentor.
 #3823 this is the pr .

umak1106 committed:
Update pkdtree.py

umak1106 commented:
> > Making those changes in pkdtree raises errors like unresolved reference to self
> 
> Can you paste the full error message please? It is difficult to figure out what can go wrong without it.

It got fixed . There are no mypy errors in this module now .

jbarnoud reviewed (approved):
Youhou! :tada:

jbarnoud committed:
Merge branch 'develop' into patch-2

jbarnoud commented:
Congratulations on that PR being merged!

umak1106 commented:
> Congratulations on that PR being merged!

Thanks
